### PR TITLE
lib: location: refactor GNSS assistance data need handling

### PR DESCRIPTION
**Pull request by tokangas**
Added support for querying GNSS assistance data need using the `nrf_modem_gnss_agps_expiry_get()` function. `NRF_MODEM_GNSS_EVT_AGPS_REQ` event is also still supported to maintain compatibility with older MFWs.

Refactored the execution order of assistance related functions. Earlier the code was splitted in several different work items, which were not executed in the logical order in the work queue. Because of this A-GPS data download didn't start before the RRC connection was released, causing unnecessary delay and RRC connection re-establishment.

Earlier filtered ephemerides were always enabled by the library when `CONFIG_NRF_CLOUD_AGPS_FILTERED_RUNTIME` was selected. Because the library does not have an API for filtered ephemerides runtime configuration, it makes more sense to not use filtered ephemerides in this case. Filtered ephemerides are only used when `CONFIG_NRF_CLOUD_AGPS_FILTERED` is selected and `CONFIG_NRF_CLOUD_AGPS_FILTERED_RUNTIME` not selected.

### CODE REVIEW

**Comment by tokangas** [comment]
I created this PR as a draft, because further logic may be needed to restrict the amount of assistance data requests from the library. Currently, assistance is requested every time the GNSS method is started and modem indicates need for assistance.

**Comment by tokangas** [comment]
I created PR #9703 which contains the changes not related to taking the `nrf_modem_gnss_agps_expiry_get()` API into use. I will rebase this once the other PR gets merged.

File: lib/location/method_gnss.c [hunk 1]
```diff
@@ -412,16 +422,9 @@ bool method_gnss_agps_required(struct nrf_modem_gnss_agps_data_frame *request)
 #endif
 
 #if defined(CONFIG_NRF_CLOUD_AGPS) || defined(CONFIG_NRF_CLOUD_PGPS)
-static void method_gnss_request_assistance(void)
+static void method_gnss_assistance_request(void)
```
Comment (trantanen) [inline lib/location/method_gnss.c:425]: This may not be the correct place but I think we need to have a pseudo-algorithm on how A-GPS (and maybe P-GPS) data retrieval works with all throttling, LTE connection status, RRC connection release, PSM etc. Now it's documented but scattered around this file so it's hard to follow. This could be documented somewhere in this file (a single place whether it's perfect or not), library documentation (maybe too detailed stuff in there), or somewhere else internally. We can discuss about this.

File: lib/location/method_gnss.c [hunk 2]
```diff
@@ -57,8 +57,21 @@ BUILD_ASSERT(
 #define AT_MDM_SLEEP_NOTIF_START "AT%%XMODEMSLEEP=1,%d,%d"
 #endif
 #if (defined(CONFIG_NRF_CLOUD_AGPS) || defined(CONFIG_NRF_CLOUD_PGPS))
+static int64_t agps_req_timestamp;
```
Comment (trantanen) [inline lib/location/method_gnss.c:60]: Static into the function where this is used? I don't know if we do that because seeing sram consumption from the beginning of the file is not accurate but otherwise it's nicer.

File: lib/location/method_gnss.c (hunk 1 above)
Comment (tokangas) [reply lib/location/method_gnss.c:425]: Yes, that's probably needed. I'm currently running tests with A-GPS + P-GPS and I'm already thinking about improving the logic with P-GPS. So, it's probably going to get even more complicated. Not sure if we have other potential places than here in the C file or in the library documentation.

File: lib/location/method_gnss.c (hunk 2 above)
Comment (tokangas) [reply lib/location/method_gnss.c:60]: Maybe it's better inside the function, because it's only used in a single place.

Comment (tokangas) [reply lib/location/method_gnss.c]: I modified the implementation and `agps_req_timestamp` is now accessed from two different places, so need to have it in the beginning of the file.

File: lib/location/method_gnss.c [hunk 3]
```diff
@@ -370,93 +391,126 @@ static void method_gnss_pgps_request_work_fn(struct k_work *item)
...
-			&method_gnss_agps_request_work);
+		method_gnss_nrf_cloud_agps_request();
 #endif
 	}
-#endif /* CONFIG_NRF_CLOUD_AGPS */
 #if defined(CONFIG_NRF_CLOUD_PGPS)
+	else if (pgps_agps_request.sv_mask_ephe != 0) {
+		/* When both A-GPS and P-GPS are enabled, this needs to be called only when no
+		 * A-GPS data is requested. When A-GPS data is requested, the nRF Cloud library
+		 * calls the function internally after processing the A-GPS data.
```
Comment (jhirsi) [inline lib/location/method_gnss.c:478]: Nasty dependency to nRF Cloud functionality. How does this work with EXTERNAL features? e.g. ATv2 is calling for nrf_cloud_pgps_notify_prediction()

File: lib/location/method_gnss.c [hunk 4]
```diff
@@ -370,93 +391,126 @@ static void method_gnss_pgps_request_work_fn(struct k_work *item)
...
+		enum lte_lc_nw_reg_status reg_status = LTE_LC_NW_REG_NOT_REGISTERED;
+
+		lte_lc_nw_reg_status_get(&reg_status);
+		if (reg_status == LTE_LC_NW_REG_REGISTERED_HOME ||
+		    reg_status == LTE_LC_NW_REG_REGISTERED_ROAMING) {
+			/* Only store the timestamp when LTE is connected, otherwise it's not
+			 * very likely that the A-GPS request succeeds.
+			 */
+			agps_req_timestamp = k_uptime_get();
+		}
```
Comment (jhirsi) [inline lib/location/method_gnss.c:467]: else: should we set agps_req_timestamp to zero?

File: lib/location/method_gnss.c [hunk 5]
```diff
@@ -828,13 +879,103 @@ static void method_gnss_positioning_work_fn(struct k_work *work)
...
+				"firmware");
+		} else {
+			LOG_WRN("nrf_modem_gnss_agps_expiry_get() failed, err %d", err);
+		}
+
+		/* Start and stop GNSS to trigger NRF_MODEM_GNSS_EVT_AGPS_REQ event if assistance
+		 * data is needed.
+		 */
+		nrf_modem_gnss_start();
+		nrf_modem_gnss_stop();
```
Comment (jhirsi) [inline lib/location/method_gnss.c:969]: Do we want to do these also in other than NRF_EOPNOTSUPP error cases? ENOMEM/ESHUTDOWN.

File: lib/location/method_gnss.c (hunk 1 above)
Comment (jhirsi) [reply lib/location/method_gnss.c:419]: Yes, I agree. pseudoalg would be appreciated. Also, there are a lot of compile time ifdefs and was thinking if all configuration options are documented? I.e what could be compiled on at the same time? AGPS/PGPS_EXTERNAL cannot be configured on at the same time as CONFIG_NRF_CLOUD_PGPS/CONFIG_NRF_CLOUD_AGPS. SO, wondering should we have also build assert for those?

**Review by jhirsi — COMMENTED** [review]
Looks very good changes for me! Some comments/questions.

File: lib/location/method_gnss.c (hunk 3 above)
Comment (tokangas) [reply lib/location/method_gnss.c:478]: Yes, it took me some time to track down why P-GPS injected the same predictions several times in a row. I think the reasoning is that the application can't do this, because it may not anymore know when A-GPS data has been received and injected.

External works in the same way, it uses the same nRF Cloud library A-GPS and P-GPS implementation.

File: lib/location/method_gnss.c (hunk 4 above)
Comment (tokangas) [reply lib/location/method_gnss.c:467]: I don't think we ever want to zero the timestamp, it can always keep the timestamp of the last A-GPS request. If execution gets to this point, the `agps_req_timestamp` is either still zero or more than one hour has passed since the previous A-GPS request. In neither case the timestamp restricts sending of an A-GPS request.

File: lib/location/method_gnss.c (hunk 3 above)
Comment (jhirsi) [reply lib/location/method_gnss.c:478]: ok, and we cannot run this code here with external?

File: lib/location/method_gnss.c (hunk 5 above)
Comment (tokangas) [reply lib/location/method_gnss.c:969]: That's probably not necessary. If the n`rf_modem_gnss_agps_expiry_get()` fails with `-NRF_ENOMEM` or `-NRF_ESHUTDOWN`, start/stop isn't most likely going to succeed either. I will change the implementation so, that start/stop is only done in case `-NRF_EOPNOTSUPP` is returned.

File: lib/location/method_gnss.c (hunk 1 above)
Comment (tokangas) [reply lib/location/method_gnss.c:419]: Actually, `CONFIG_NRF_CLOUD_AGPS` and `CONFIG_NRF_CLOUD_PGPS` are always needed for A-GPS and P-GPS support. The same nRF Cloud library implementation is used also with `CONFIG_LOCATION_METHOD_GNSS_AGPS_EXTERNAL` and `CONFIG_LOCATION_METHOD_GNSS_PGPS_EXTERNAL`.

I'll come up with some kind of pseudo code/algorithm for the logic, because it's not that obvious how it all works.

File: lib/location/method_gnss.c (hunk 3 above)
Comment (tokangas) [reply lib/location/method_gnss.c:478]: That code is used also with external, right? 😄

File: lib/location/method_gnss.c (hunk 1 above)
Comment (jhirsi) [reply lib/location/method_gnss.c:419]: OK, thanks

File: lib/location/method_gnss.c (hunk 3 above)
Comment (jhirsi) [reply lib/location/method_gnss.c:478]: yes, that how it looked. So, I was confused why ATv2 (which is using location api external config) is also calling for nrf_cloud_pgps_notify_prediction().

Comment (tokangas) [reply lib/location/method_gnss.c:478]: The ATv2 P-GPS implementation should be checked, there's probably some fine tuning to be done there as well.

Comment (trantanen) [reply lib/location/method_gnss.c:478]: ATv2 P-GPS is not fully cleaned after taking Location library into use and there's a ticket to check exactly these "duplicate" actions, which I didn't dare to remove originally.

**Review by jhirsi — APPROVED** [review]

**Comment by tokangas** [comment]
@moksanen This PR breaks our tests, because one of the A-GPS debug traces in the Location library has changed, "A-GPS request from modem" is now just "A-GPS request". I changed the trace, because the printed data is not the request we get from the modem, but what the library actually decides to request using A-GPS. These two may or may not be the same, it depends on the configuration.

File: lib/location/method_gnss.c [hunk 6]
```diff
@@ -828,13 +906,135 @@ static void method_gnss_positioning_work_fn(struct k_work *work)
...
+		}
+
+		return;
+	}
+
+	agps_expiry_get_supported = true;
+
+	if (method_gnss_agps_expiry_process(&agps_expiry)) {
+		method_gnss_assistance_request();
+	}
```
Comment (trantanen) [inline lib/location/method_gnss.c:1037]: Cannot use proper line but check whether starting and stopping GNSS is valid in comments on line 1083:
```
	 * downloaded before GNSS is started. GNSS is briefly started and stopped to trigger
	 * the NRF_MODEM_GNSS_EVT_AGPS_REQ event, which in turn causes the A-GPS data download
```

**Review by trantanen — COMMENTED** [review]
Good changes overall.
I was thinking that some of the comments could even be in the documentation but I guess it's good for now to have things documented at least in the function comments.

File: lib/location/method_gnss.c [hunk 7]
```diff
@@ -370,93 +391,153 @@ static void method_gnss_pgps_request_work_fn(struct k_work *item)
...
+ * There are three possible assistance configurations with different behavior:
+ *
+ * A-GPS only:
+ * Ephemerides, almanacs and additional assistance data is handled by A-GPS.
+ * If any additional assistance data (UTC, Klobuchar, GPS time, integrity or position) is needed,
+ * all additional assistance is requested at the same time.
+ *
+ * A-GPS and P-GPS:
+ * Ephemerides are handled by P-GPS.
+ * Almanacs are not used.
```
[the remark this answers is not in the document]
Comment (tokangas) [reply lib/location/method_gnss.c:433]: Almanacs are like ephemerides, but less accurate (and valid for a much longer time). With P-GPS the assumption is that valid ephemerides are always available, so there's no need to have almanacs as well.

When ephemerides are not available, almanacs are used to determine which satellites are above the horizon and to predict dopplers for each satellite, making satellite acquisition faster.

File: lib/location/method_gnss.c [hunk 8]
```diff
@@ -370,93 +391,153 @@ static void method_gnss_pgps_request_work_fn(struct k_work *item)
...
-		type_count++;
-	}
-#endif
-	if (request->data_flags & NRF_MODEM_GNSS_AGPS_GPS_UTC_REQUEST) {
-		type_count++;
-	}
-	if (request->data_flags & NRF_MODEM_GNSS_AGPS_KLOBUCHAR_REQUEST) {
-		type_count++;
-	}
-	if (request->data_flags & NRF_MODEM_GNSS_AGPS_NEQUICK_REQUEST) {
```
[the remark this answers is not in the document]
Comment (tokangas) [reply lib/location/method_gnss.c:392]: Klobuchar and NeQuick are alternative models for ionospheric corrections. Although NeQuick is present in the GNSS API, we currently only have support for Klobuchar and there are no plans to add support for NeQuick in the future. While NeQuick would allow more accurate corrections, it's much more resource intensive than Klobuchar.

The reason why I removed it here is that it would not have been present in `method_gnss_assistance_request()` where "all" data_flags bits are set, possible causing confusion why NeQuick is missing there.

That said, maybe the GNSS API should have a note saying that NeQuick corrections are not supported by GNSS. Or, I guess we could even remove NeQuick from the API, if we don't plan on adding support for it later.

File: lib/location/method_gnss.c (hunk 6 above)
Comment (tokangas) [reply lib/location/method_gnss.c:1037]: Yes, it's valid if `nrf_modem_gnss_agps_expiry_get()` is not supported. I'll mention this in the comment for clarity.

Comment (tokangas) [reply lib/location/method_gnss.c:1037]: It came to my mind, that we have a flag for `nrf_modem_gnss_agps_expiry_get()` support. Thus, the sleep could be done only when needed. However, the flag is set only when `method_gnss_agps_req_work` is executed for the first time, so it has not yet been set when execution gets here for the first time. Maybe we can just always sleep for 100 ms here. 😄

Comment (trantanen) [reply lib/location/method_gnss.c:1040]: Yep, sleeping 100ms always is fine

File: lib/location/method_gnss.c (hunk 8 above)
[the remark this answers is not in the document]
Comment (trantanen) [reply lib/location/method_gnss.c:392]: > I guess we could even remove NeQuick from the API, if we don't plan on adding support for it later.

I'd go for this after checking if we have any plans for supporting it

File: lib/location/method_gnss.c (hunk 7 above)
[the remark this answers is not in the document]
Comment (trantanen) [reply lib/location/method_gnss.c:436]: Something along the lines of your first paragraph would be nice to see written down somewhere in Location library. Maybe just into this comment block.

File: lib/location/method_gnss.c (hunk 8 above)
[the remark this answers is not in the document]
Comment (tokangas) [reply lib/location/method_gnss.c:392]: I created a Jira ticket for this.

File: lib/location/method_gnss.c (hunk 7 above)
[the remark this answers is not in the document]
Comment (tokangas) [reply lib/location/method_gnss.c:436]: Added a couple of lines for the almanacs.

**Review by trantanen — APPROVED** [review]

File: lib/location/method_gnss.c [hunk 9]
```diff
@@ -828,13 +912,135 @@ static void method_gnss_positioning_work_fn(struct k_work *work)
...
+ *
+ * Almanacs:
+ * Each satellite has its own almanac expiration time. The code goes though all satellites and
+ * checks whether its almanac has expired or is going to expire. If there are enough satellites
+ * with expired almanacs, almanacs for all satellites are requested.
+ *
+ * Other assistance data:
+ * For other assistance data, the same A-GPS expiration threshold is used if expiry time is
+ * provided by GNSS. For some data, there's only a flag indicating whether the data is needed or
+ * not.
```
Comment (peknis) [inline lib/location/method_gnss.c:938]:
```suggestion
 * provided by GNSS. For some data, there is only a flag indicating whether the data is needed.
```

File: lib/location/method_gnss.c [hunk 10]
```diff
@@ -67,12 +88,18 @@ BUILD_ASSERT(
 static struct k_work method_gnss_start_work;
 static struct k_work method_gnss_pvt_work;
 #if defined(CONFIG_NRF_CLOUD_AGPS) || defined(CONFIG_NRF_CLOUD_PGPS)
+/* Flag indicating whether nrf_modem_gnss_agps_expiry_get() is supported or not. If the API is
```
Comment (peknis) [inline lib/location/method_gnss.c:91]:
```suggestion
/* Flag indicating whether nrf_modem_gnss_agps_expiry_get() is supported. If the API is
```

**Review by peknis — CHANGES_REQUESTED** [review]

File: lib/location/method_gnss.c [hunk 11]
```diff
@@ -59,6 +59,27 @@ BUILD_ASSERT(
...
+ * predictions.
+ */
+#define PGPS_EXPIRY_THRESHOLD 0
+/* A-GPS minimum number of expired ephemerides to request all ephemerides. */
+#define AGPS_EPHE_MIN_COUNT 3
+/* P-GPS minimum number of expired ephemerides to trigger injection of predictions. With P-GPS
+ * predictions are not available for all satellites, especially when predictions are made further
+ * into the future, so it's natural that predictions are missing for some of the satellites.
+ */
+#define PGPS_EPHE_MIN_COUNT 12
```
[the remark this answers is not in the document]
Comment (tokangas) [reply lib/location/method_gnss.c:80]: I selected this number so, that it would be higher than the number of ephes possibly missing in the predictions. The idea is, that new predictions do not need to be injected if GNSS already has the correct set (which usually does not have 32 ephemerides). Calling `nrf_cloud_pgps_notify_prediction()` would only cause the same set of ephemerides to be injected again.

The P-GPS library handles injection of the next set of predictions automatically with a timer. It also triggers download for new predictions when all previous predictions have been used. However, I realized now, that pre-emptive downloads do not probably work. Usually, `nrf_cloud_pgps_preemptive_updates()` has been called after injecting ephemerides, but the time based injection happens internally in the P-GPS library, so pre-emptive updates do not get triggered. I need to see where we should check if pre-emptive P-GPS updates are needed.

File: lib/location/method_gnss.c [hunk 12]
```diff
@@ -370,93 +394,156 @@ static void method_gnss_pgps_request_work_fn(struct k_work *item)
...
+ * P-GPS only:
+ * Ephemerides are handled by P-GPS.
+ * Almanacs are not used.
+ * GPS time and position assistance are handled by P-GPS.
+ *
+ * The frequency of A-GPS data requests is limited to avoid requesting data again and again for
+ * example in case the server is down.
+ *
+ * The reason why almanacs are not used with P-GPS is, that GNSS always has valid (and more
+ * accurate) ephemerides available. Thus, there's no benefit from having almanacs.
```
[the remark this answers is not in the document]
Comment (tokangas) [reply lib/location/method_gnss.c:450]: As discussed offline, I'll change the implementation so, that almanacs are downloaded in case predictions are downloaded for a period longer than one week. This way almanacs can be used at least to help satellite acquisition if predicted ephemerides are not available.

File: lib/location/method_gnss.c (hunk 11 above)
[the remark this answers is not in the document]
Comment (tokangas) [reply lib/location/method_gnss.c:81]: Pre-emptive P-GPS updates are now checked when a GNSS fix is got, timeout is triggered or the location request is cancelled. I removed the check from prediction injection, because I don't think we need both.

**Review by peknis — APPROVED** [review]

**Review by hiltunent — APPROVED** [review]

File: lib/location/method_gnss.c [hunk 13]
```diff
@@ -370,93 +393,162 @@ static void method_gnss_pgps_request_work_fn(struct k_work *item)
...
+ * Ephemerides are handled by P-GPS.
+ * Almanacs are handled by A-GPS, but only if the downloaded P-GPS prediction set is longer than
+ * one week.
+ * Additional assistance data is handled by A-GPS.
+ * If any additional assistance data (UTC, Klobuchar, GPS time, integrity or position) is needed,
+ * all additional assistance is requested at the same time.
+ *
+ * P-GPS only:
+ * Ephemerides are handled by P-GPS.
+ * Almanacs are not used.
```
Comment (hiltunent) [inline lib/location/method_gnss.c:443]: maybe remove this line?
